src/sinks/datadog_archives.rs

```diff
@@ -0,0 +1,585 @@
+            })
+            .settings(request, S3RetryLogic)
+            .service(s3);
+        Ok(svc)
+    }
+}
+
+const RESERVED_ATTRIBUTES: [&'static str; 10] = [
+    "_id", "date", "message", "host", "source", "service", "status", "tags", "trace_id", "span_id",
+];
+
+struct DatadogArchivesSinkEncoding {
+    key_prefix_template: Template,
+    reserved_attributes: HashSet<&'static str>,
+    host_number: Vec<u8>,
+    seq_number: AtomicU32,
+}
+
+impl DatadogArchivesSinkEncoding {
+    /// Applies the following transformations to align event's schema with DD:
+    /// - `_id` is generated in the sink(format described below);
+    /// - `date` is set from the Global Log Schema's `timestamp` mapping, or to the current time if missing;
+    /// - `message`,`host` are set from the corresponding Global Log Schema mappings;
+    /// - `source`, `service`, `status`, `tags` and other reserved attributes are left as is;
+    /// - the rest of the fields is moved to `attributes`.
+    fn encode_event(
+        &mut self,
+        event: Event,
+    ) -> Option<EncodedEvent<PartitionInnerBuffer<Vec<u8>, Bytes>>> {
+        let key = self
+            .key_prefix_template
+            .render_string(&event)
+            .map_err(|error| {
+                emit!(TemplateRenderingFailed {
+                    error,
+                    field: Some("key_prefix"),
+                    drop_event: true,
+                });
+            })
+            .ok()?;
+
+        let mut log = event.into_log();
+
+        log.insert("_id", self.generate_log_id());
+
+        let timestamp = log
+            .remove(crate::config::log_schema().timestamp_key())
+            .unwrap_or_else(|| chrono::Utc::now().timestamp_millis().into());
+        log.insert(
+            "date",
+            timestamp
+                .as_timestamp()
+                .cloned()
+                .unwrap_or_else(chrono::Utc::now)
+                .to_rfc3339_opts(SecondsFormat::Millis, true),
+        );
+        log.rename_key_flat(crate::config::log_schema().message_key(), "message");
+        log.rename_key_flat(crate::config::log_schema().host_key(), "host");
+
+        let mut attributes = BTreeMap::new();
+        let custom_attributes: Vec<String> = log
+            .keys()
+            .filter(|path| !self.reserved_attributes.contains(path.as_str()))
+            .collect();
+        for path in custom_attributes {
+            if let Some(value) = log.remove(&path) {
+                attributes.insert(path, value);
+            }
+        }
+        log.insert("attributes", attributes);
+
+        let mut bytes =
+            serde_json::to_vec(&log).expect("Failed to encode event as json, this is a bug!");
+        bytes.push(b'\n');
+
+        Some(EncodedEvent {
+            item: PartitionInnerBuffer::new(bytes, key.into()),
+            finalizers: log.metadata_mut().take_finalizers(),
+        })
+    }
+
+    /// Generates a unique ID compatible with DD
```

Comment:
again, there is no real public documentation I'm afraid(but can share with you some internal docs if needed)- this is not exposed in any public DD API, it's rather internals of DD Logs backend.
>Why?
Because the whole ID starts with the timestamp in millis, so in 1 millisecond an identical `seq_number`'s will not lead to ID collisions.

Reply:
Gotcha. Alright, well, please add documentation to that effect that we're relying on internal knowledge of the relevant system etc etc.